### PR TITLE
Reduce

### PR DESCRIPTION
**Pull request by skyfarliu**

### CODE REVIEW

File: view.go [hunk 1]
```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+	qb.AddIfNotZero(` AND doc_id <= $1`, p.EndKeyDocID)
+
+	qb.Add(" ORDER BY " + orderBy)
+	qb.AddIfNotZero(" LIMIT $1", p.Limit)
+	qb.AddIfNotZero(" OFFSET $1", p.Skip)
+
+	rows, err := qb.Query(ctx, v.sqldb)
+	if err != nil {
+		return ret, err
+	}
```
Comment (bigdrum) [inline view.go:263]: defer rows.Close()

File: view.go [hunk 2]
```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+	for rows.Next() {
+		r := ViewResultRow{}
+		if err := rows.Scan(&r.Key, &r.ID, &r.Value); err != nil {
+			return ret, err
+		}
+		ret.Rows = append(ret.Rows, r)
+	}
+
```
Comment (bigdrum) [inline view.go]: rows.Close()

File: view.go [hunk 3]
```diff
@@ -20,8 +24,10 @@ type ViewConfig struct {
 	Name    string
 	Version string
 
-	Inputs []ViewInput
-	Mapper func(doc *Document, emit func(ve *ViewEntry) error) error
+	// REVIEW: change this to ViewInput instead of []ViewInput? Since CouchDB views cannot query accross databases?
```
Comment (bigdrum) [inline view.go]: right. but we want to support that later. see the design doc.

Comment (bigdrum) [reply view.go]: you can leave it as a single input for now.

File: view.go [hunk 4]
```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+			return r, err
+		}
+		r.UpdateSeq = lastSeq
+	}
+
+	if p.NoReduce || v.config.Reducer == nil {
+		for i := range r.Rows {
+			if p.IncludeDocs {
+				r.Rows[i].Doc = &Document{}
+				err = v.config.Input.Table.Get(ctx, r.Rows[i].ID, r.Rows[i].Doc)
```
Comment (bigdrum) [inline view.go]: Put a TODO to perform batch fetch here.

File: view.go [hunk 5]
```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+		v.goRefresh(ctx)
+	}
+
+	return r, nil
+}
+
+// goRefresh calls Refresh in a new goroutine.
+func (v *View) goRefresh(ctx context.Context) {
+	go func() {
+		newCtx, cancel := context.WithTimeout(ctx, refreshTimeout)
```
Comment (bigdrum) [inline view.go]: i don't think we need refreshTimeout. Also add a TODO to avoid multiple ongoing refresh.

File: view_test.go [hunk 6]
```diff
@@ -108,3 +111,428 @@ func TestView(t *testing.T) {
...
+    "rows": [
+        {
+            "value": 2
+        }
+    ]
+}
+`)
+}
+
+func printJSON(i interface{}) {
```
Comment (bigdrum) [inline view_test.go]: probably put the test helper functions in da_test.go

File: view_test.go [hunk 7]
```diff
@@ -108,3 +111,428 @@ func TestView(t *testing.T) {
...
+			value, err := json.Marshal(len(d["tags"].([]interface{})))
+			if err != nil {
+				return err
+			}
+			return emit(&da.ViewEntry{
+				Key:   d["title"].(string),
+				Value: json.RawMessage(value),
+			})
+		},
+		Reducer: func(entries []da.ViewResultRow, rereduce bool) (json.RawMessage, error) {
```
Comment (bigdrum) [inline view_test.go]: the parameters should be key, values, rereduce.

File: view.go [hunk 8]
```diff
@@ -179,6 +220,137 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+				err = v.config.Input.Table.Get(ctx, r.Rows[i].ID, r.Rows[i].Doc)
+				if err != nil {
+					return r, err
+				}
+			}
+		}
+		return r, nil
+	}
+
+	value, err := v.config.Reducer(r.Rows, false)
```
Comment (bigdrum) [inline view.go]: This doesn't seem to handle querying multiple keys yet? You should panic if that's the case, and put a TODO. And the next step should implement such support.

File: view.go (hunk 3 above)
Comment (skyfarliu) [reply view.go]: done

File: view.go (hunk 1 above)
Comment (skyfarliu) [reply view.go:263]: done

File: view.go [hunk 9]
```diff
@@ -20,8 +21,11 @@ type ViewConfig struct {
 	Name    string
 	Version string
 
-	Inputs []ViewInput
+	// TODO: Use []Input to support multi-table mapreduce
+	Input  ViewInput
 	Mapper func(doc *Document, emit func(ve *ViewEntry) error) error
+	// TODO: support multiple keys.
+	Reducer func(keys []ViewReduceKey, values []interface{}, rereduce bool) (interface{}, error)
```
Comment (bigdrum) [inline view.go:28]: A single reducer only handles a single key.

File: view.go [hunk 10]
```diff
@@ -179,6 +213,152 @@ func (v *View) Read(ctx context.Context, key string, each func(entry *ViewEntry)
...
+	}
+
+	keys := make([]ViewReduceKey, len(r.Rows))
+	values := make([]interface{}, len(r.Rows))
+	for i := range r.Rows {
+		keys[i] = ViewReduceKey{Key: r.Rows[i].Key, DocID: r.Rows[i].ID}
+		values[i] = r.Rows[i].Value
+	}
+
+	value, err := v.config.Reducer(keys, values, false)
```
Comment (bigdrum) [inline view.go]: This is not how to reduce multiple keys. You should group the map output by keys, and for each group let reducer to handle it.

File: view.go [hunk 11]
```diff
@@ -31,10 +35,55 @@ type ViewInput struct {
 
 // ViewEntry represents a single entry of the view mapper output.
 type ViewEntry struct {
-	Key   string
+	Key   interface{}
```
Comment (bigdrum) [inline view.go]: I think if it is json, using json.RawMessage is better for handling read.

**Pull request merged**